File: nim/src/camera.nim
import std/math
import color
import point3
import vec3
import ray

type
  Camera* = object
    aspect, height, width, focalLength, lensRadius: float
    origin: Point3
    horizontal, vertical, lowerLeftCorner, u, v, w: Vec3

  RefCamera = ref Camera

func newCamera*(
    verticalFOV, aspect, focalLength, aperture, focusDistance: float;
    lookFrom, lookAt, vUp: Point3
  ): Camera =
  # let theta = verticalFOV * PI / 180.0
  let height = 2.0 * tan(verticalFOV / 2.0)
  let width = aspect * height

  let w = (lookFrom - lookAt).unit()
  let u = vUp.cross(w).unit()
  let v = w.cross(u)

  let horizontal = focusDistance * width * u;
  let vertical = focusDistance * height * v;

  Camera(
    aspect: aspect,
    height: height,
    width: width,
    focalLength: focalLength,
    lensRadius: aperture / 2.0,
    origin: lookFrom,
    horizontal: horizontal,
    vertical: vertical,
    w: w,
    u: u,
    v: v,
    lowerLeftCorner: lookFrom - horizontal / 2.0 - vertical / 2.0 - focusDistance * w
  )

func newRefCamera*(
    verticalFOV, aspect, focalLength, aperture, focusDistance: float;
    lookFrom, lookAt, vUp: Point3
  ): RefCamera =
  # let theta = verticalFOV * PI / 180.0
  let height = 2.0 * tan(verticalFOV / 2.0)
  let width = aspect * height

  let w = (lookFrom - lookAt).unit()
  let u = vUp.cross(w).unit()
  let v = w.cross(u)

  let horizontal = focusDistance * width * u;
  let vertical = focusDistance * height * v;

  RefCamera(
    aspect: aspect,
    height: height,
    width: width,
    focalLength: focalLength,
    lensRadius: aperture / 2.0,
    origin: lookFrom,
    horizontal: horizontal,
    vertical: vertical,
    w: w,
    u: u,
    v: v,
    lowerLeftCorner: lookFrom - horizontal / 2.0 - vertical / 2.0 - focusDistance * w
  )

proc getRay*(self: Camera; s, t: float): Ray =
  let rd = self.lensRadius * randomPointInUnitDisc()
  let offset = self.u * rd.x + self.v * rd.y
  return Ray(
    origin: self.origin + offset,
    direction: self.lowerLeftCorner + s * self.horizontal + t * self.vertical - self.origin - offset,
    color: Color(x: 1.0, y: 1.0, z: 1.0)
  )

File: rust/src/camera.rs
use crate::color::Color;
use crate::point3::Point3;
use crate::ray::Ray;
use crate::vec3::Vec3;

pub struct Camera {
    pub aspect: f32,
    pub height: f32,
    pub width: f32,
    pub focal_length: f32,
    pub origin: Point3,
    pub horizontal: Vec3,
    pub vertical: Vec3,
    pub lower_left_corner: Vec3,
}

impl Camera {
    pub fn new(vertical_fov_rad: f32, aspect: f32, focal_length: f32, origin: Point3) -> Camera {
        let height = 2.0 * (vertical_fov_rad / 2.0).tan();
        let width = aspect * height;
        let horizontal = Vec3 {
            x: width,
            y: 0.0,
            z: 0.0,
        };
        let vertical = Vec3 {
            x: 0.0,
            y: height,
            z: 0.0,
        };
        Camera {
            aspect,
            height,
            width,
            focal_length,
            origin,
            horizontal,
            vertical,
            lower_left_corner: origin
                - horizontal / 2.0
                - vertical / 2.0
                - Vec3 {
                    x: 0.0,
                    y: 0.0,
                    z: focal_length,
                },
        }
    }

    pub fn get_ray(&self, u: f32, v: f32) -> Ray {
        Ray {
            origin: self.origin,
            direction: self.lower_left_corner + u * self.horizontal + v * self.vertical
                - self.origin,
            color: Color::white(),
        }
    }
}
